machineconfiguration/v1alpha1/types_machineosconfig.go

```diff
@@ -0,0 +1,203 @@
+	// +kubebuilder:default:=Default
+	// +kubebuilder:validation:Enum:=Default;OpenShiftImageBuilder;PodImageBuilder
+	// Valid options are: OpenShiftImageBuilder, PodImageBuilder, and Default (OpenShiftImageBuilder)
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+	// baseOSExtensionsImageURL is the base Extensions image used in the build process
+	// +kubebuilder:validation:Pattern=`^https://`
+	// +optional
+	BaseOSExtensionsImageURL string `json:"baseOSExtensionsImageURL"`
+	// baseOSImageURL is the base OSImage we use to build our custom image.
+	// +optional
+	BaseOSImageURL string `json:"baseOSImageURL"`
+	// baseImagePullSecret is the secret used to pull the base image.
+	// +kubebuilder:validation:Required
+	BaseImagePullSecret ImageSecretObjectReference `json:"baseImagePullSecret"`
+	// finalImagePushSecret is the secret used to connect to a user registry.
+	// +kubebuilder:validation:Required
+	FinalImagePushSecret ImageSecretObjectReference `json:"finalImagePushSecret"`
+	// finalImagePullSecret is the secret used to pull the final produced image.
+	// +optional
+	FinalImagePullSecret ImageSecretObjectReference `json:"finalImagePullSecret"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// +listType=set
+	// +kubebuilder:validation:MinItems=0
+	// +kubebuilder:validation:MaxItems=2
+	// +optional
+	Containerfile []string `json:"containerFile"`
+	// finalImagePullspec describes the location of the final image.
+	// +kubebuilder:validation:Required
+	FinalImagePullspec string `json:"finalImagePullspec"`
+}
+
+// Refers to the name of a MachineConfigPool (e.g., "worker", "infra", etc.):
+type MachineConfigPoolReference struct {
+	// name of the MachineConfigPool object.
+	// +kubebuilder:validation:MaxLength:=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of a rendered MachineConfig (e.g., "rendered-worker-ec40d2965ff81bce7cd7a7e82a680739", etc.):
+type RenderedMachineConfigReference struct {
+	// name is the name of the rendered MachineConfig object.
+	// +kubebuilder:validation:MaxLength:=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of an image registry push/pull secret needed in the build process.
+type ImageSecretObjectReference struct {
+	// name is the name of the secret used to push or pull this MachineOSConfig object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+type TriggerRebuild string
+
+const (
+	// RebuildNow indicates that the user wants to rebuild their image immediately
+	RebuildNow TriggerRebuild = "RebuildNow"
```

Comment:
will this cause an infinite rebuild loop?

Reply:
I will wire up logic in the MCO so that if he user sets `RebuildNow` or `RebuildDelayed`, the next rebuild will cause this to get get back to `NoRebuild`

Reply:
What is the mechanism for defining the delay interval for `RebuildDelayed`

Reply:
Delayed will kick off a build when the current build (if there is one) either succeeds or fails. I will add a godoc detailing this

Reply:
> I will wire up logic in the MCO so that if he user sets `RebuildNow` or `RebuildDelayed`, the next rebuild will cause this to get get back to `NoRebuild`

this is bad practice.  Fields should be owned by exactly one entity, not shared between users and controllers.  See the ForceRedeploymentReason mechanic employed by static pod operators: https://github.com/openshift/api/blob/master/operator/v1/types.go#L191